File: src/class-problems/valid-parentheses/20.js
// 20. Valid Parentheses

/**
 * @param {string} s
 * @returns {boolean}
 */
const isValid = (str) => {
  const stacks = [];
  const brackets = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  for (char of str) {
    if (Object.values(brackets).includes(char)) {
      stacks.push(char);
    } else if (brackets[char] !== stacks.pop()) {
      return false;
    }
  }
  return stacks.length === 0;
};

File: src/leetcode/valid-parentheses/20.py
#20. Valid Parentheses

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        stack = []
        brackets = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in brackets.values():
                stack.append(char)
            else:
                if not stack or stack.pop() != brackets[char]:
                    return False

        return not stack